File: neetcode/543. Diameter of Binary Tree/543.java
/*
Method 1.0:
Intuition: Use DFS, get left and right subtree heights to calculate curr diameter,
update diameter with max value. DFS returns the height of the subtree with the curr node
as its root.
TC: O(N), visiting N nodes
SC: O(log N), recursion stack space
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int diameter;

    public int diameterOfBinaryTree(TreeNode root) {
        diameter = 0;
        findHeight(root);
        return diameter;
    }

    private int findHeight(TreeNode root) {
        if (root == null) return 0;

        int leftHeight = findHeight(root.left);
        int rightHeight = findHeight(root.right);

        diameter = Math.max(diameter, leftHeight + rightHeight);

        return 1 + Math.max(leftHeight, rightHeight);
    }
}

File: neetcode/543. Diameter of Binary Tree/543.kt
/*
Method 1.0:
Intuition: Use DFS, get left and right subtree heights to calculate curr diameter,
update diameter with max value. DFS returns the height of the subtree with the curr node
as its root.
TC: O(N), visiting N nodes
SC: O(log N), recursion stack space
 */
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    var diameter: Int = 0

    fun diameterOfBinaryTree(root: TreeNode?): Int {
        findHeight(root)
        return diameter
    }

    fun findHeight(root: TreeNode?): Int {
        if (root == null) return 0

        val leftHeight = findHeight(root.left)
        val rightHeight = findHeight(root.right)

        diameter = max(diameter, leftHeight + rightHeight)

        return 1 + max(leftHeight, rightHeight)
    }
}
